"Duplicate" scroll events on some mice

I have this in my `rc.xml` to change workspaces using the scroll wheel:
```
        <context name="Root">
            <mousebind direction="Up" action="Scroll">
                <action name="GoToDesktop" to="left" wrap="no" />
            </mousebind>
            <mousebind direction="Down" action="Scroll">
                <action name="GoToDesktop" to="right" wrap="no" />
            </mousebind>
        </context>
```
Works fine with my old Logitech M500.

With my Logitech M705, though, labwc jumps multiple workspace at once during one "click" of the wheel. (By "click" I mean the tactile feedback while turning the wheel -- I'm not pressing down the wheel, like I would to issue a middle mouse click).

I had a look at the output of `wev`, but I can't make heads or tails of it.

M500 (ok) -- one click of the wheel results in a "frame" line and the three "axis" lines:
```
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6172381; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6172861; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6173341; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6173949; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6174517; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6175077; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6175669; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6176365; axis: 0 (vertical), value: -15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6176397; axis: 0 (vertical), value: -15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6177061; axis: 0 (vertical), value: -15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6177621; axis: 0 (vertical), value: -15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6178189; axis: 0 (vertical), value: -15.000000
```
M705 (not ok):
```
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6218674; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6219670; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: 1
[13:      wl_pointer] axis: time: 6220724; axis: 0 (vertical), value: 15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6222086; axis: 0 (vertical), value: -15.000000
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 0 (vertical), discrete: -1
[13:      wl_pointer] axis: time: 6223280; axis: 0 (vertical), value: -15.000000
```
During a physical click of the M705's scroll wheel, I see the "axis" lines. I have no idea what those "frame" lines are supposed to mean. They occur in between the clicks. If I turn the scroll wheel *very slowly* (without triggering a click), I can see the "frame" lines appear one by one.

Scrolling in applications like Firefox or l3afpad behaves the same for both mice, though.

For comparison: On plain X11, I only see button events like "button 4". It's the same for both mice.

Any ideas what this could be? (Is this even a labwc bug or should I take this up with the wlroots guys or libinput? Since applications work fine on Wayland, I suspect it has something to do with labwc ...)

Yep, here we go:

<details><summary>M705 (hires)</summary>

```
00:00:02.642 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 0.000000 and delta 7.500000, delta discrete 60.000000
00:00:02.642 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: no
--------------
00:00:02.658 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:02.658 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:02.666 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:02.666 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:02.674 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:02.674 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:02.682 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:02.682 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:02.820 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 1.875000, delta discrete 15.000000
00:00:02.820 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.875000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:05.148 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.875000 and delta 7.500000, delta discrete 60.000000
00:00:05.148 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:05.156 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:05.156 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:05.164 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 1.875000, delta discrete 15.000000
00:00:05.164 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 8.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:05.172 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 8.125000 and delta 1.875000, delta discrete 15.000000
00:00:05.172 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:07.726 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta 7.500000, delta discrete 60.000000
00:00:07.726 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:07.734 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:07.734 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:07.742 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:07.742 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:07.750 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:07.750 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:07.760 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:07.760 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:10.366 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 7.500000, delta discrete 60.000000
00:00:10.366 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:10.374 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta 1.875000, delta discrete 15.000000
00:00:10.374 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:10.390 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:10.390 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:10.406 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 1.875000, delta discrete 15.000000
00:00:10.406 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 8.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:10.492 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 8.125000 and delta 1.875000, delta discrete 15.000000
00:00:10.492 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:12.936 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta 7.500000, delta discrete 60.000000
00:00:12.936 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:12.944 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:12.944 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:12.952 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:12.952 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:12.968 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:12.968 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:12.976 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:12.976 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:15.744 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 7.500000, delta discrete 60.000000
00:00:15.744 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:15.768 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta 1.875000, delta discrete 15.000000
00:00:15.768 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:15.792 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:15.792 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:15.826 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 1.875000, delta discrete 15.000000
00:00:15.826 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 8.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:15.922 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 8.125000 and delta 1.875000, delta discrete 15.000000
00:00:15.922 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:18.334 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta 7.500000, delta discrete 60.000000
00:00:18.334 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:18.342 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:18.342 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:18.350 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:18.350 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:18.358 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:18.358 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:18.366 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:18.366 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:20.966 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 7.500000, delta discrete 60.000000
00:00:20.966 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:20.990 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta 1.875000, delta discrete 15.000000
00:00:20.990 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:20.998 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:20.998 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:21.034 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 1.875000, delta discrete 15.000000
00:00:21.034 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 8.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:21.204 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 8.125000 and delta 1.875000, delta discrete 15.000000
00:00:21.204 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:23.452 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta 7.500000, delta discrete 60.000000
00:00:23.452 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:23.476 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:23.476 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:23.484 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:23.484 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:23.500 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:23.500 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:23.594 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:23.594 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:25.932 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 7.500000, delta discrete 60.000000
00:00:25.932 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:25.942 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta 1.875000, delta discrete 15.000000
00:00:25.942 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:25.950 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:25.950 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:25.958 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 3.750000, delta discrete 30.000000
00:00:25.958 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 3.750000, delta discrete 30.000000 which should run action: no
--------------
00:00:28.480 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta 7.500000, delta discrete 60.000000
00:00:28.480 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 7.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:28.496 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 7.500000 and delta 1.875000, delta discrete 15.000000
00:00:28.496 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 9.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:28.504 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 9.375000 and delta 1.875000, delta discrete 15.000000
00:00:28.504 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 1.250000, delta 1.875000, delta discrete 15.000000 which should run action: yes
--------------
00:00:28.512 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 1.250000 and delta 1.875000, delta discrete 15.000000
00:00:28.512 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 3.125000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:28.528 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 3.125000 and delta 1.875000, delta discrete 15.000000
00:00:28.528 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 5.000000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:31.152 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 5.000000 and delta 7.500000, delta discrete 60.000000
00:00:31.152 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta 7.500000, delta discrete 60.000000 which should run action: yes
--------------
00:00:31.174 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta 1.875000, delta discrete 15.000000
00:00:31.174 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:31.182 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:31.182 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:31.190 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 3.750000, delta discrete 30.000000
00:00:31.190 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 3.750000, delta discrete 30.000000 which should run action: no
--------------
00:00:33.930 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 10.000000 and delta -7.500000, delta discrete -60.000000
00:00:33.930 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 2.500000, delta -7.500000, delta discrete -60.000000 which should run action: no
--------------
00:00:33.938 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 2.500000 and delta -1.875000, delta discrete -15.000000
00:00:33.938 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 0.625000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:33.946 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 0.625000 and delta -1.875000, delta discrete -15.000000
00:00:33.946 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -1.250000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:33.954 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -1.250000 and delta -3.750000, delta discrete -30.000000
00:00:33.954 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -5.000000, delta -3.750000, delta discrete -30.000000 which should run action: no
--------------
00:00:36.714 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -5.000000 and delta -7.500000, delta discrete -60.000000
00:00:36.714 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -2.500000, delta -7.500000, delta discrete -60.000000 which should run action: yes
--------------
00:00:36.722 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -2.500000 and delta -1.875000, delta discrete -15.000000
00:00:36.722 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -4.375000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:36.730 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -4.375000 and delta -1.875000, delta discrete -15.000000
00:00:36.730 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -6.250000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:36.738 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -6.250000 and delta -1.875000, delta discrete -15.000000
00:00:36.738 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -8.125000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:36.754 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -8.125000 and delta -1.875000, delta discrete -15.000000
00:00:36.754 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -10.000000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:39.280 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -10.000000 and delta -7.500000, delta discrete -60.000000
00:00:39.280 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -7.500000, delta -7.500000, delta discrete -60.000000 which should run action: yes
--------------
00:00:39.288 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -7.500000 and delta -1.875000, delta discrete -15.000000
00:00:39.288 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -9.375000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:39.296 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -9.375000 and delta -1.875000, delta discrete -15.000000
00:00:39.296 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -1.250000, delta -1.875000, delta discrete -15.000000 which should run action: yes
--------------
00:00:39.304 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -1.250000 and delta -1.875000, delta discrete -15.000000
00:00:39.304 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, -3.125000, delta -1.875000, delta discrete -15.000000 which should run action: no
--------------
00:00:41.988 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, -3.125000 and delta 7.500000, delta discrete 60.000000
00:00:41.988 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 4.375000, delta 7.500000, delta discrete 60.000000 which should run action: no
--------------
00:00:41.996 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 4.375000 and delta 1.875000, delta discrete 15.000000
00:00:41.996 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 6.250000, delta 1.875000, delta discrete 15.000000 which should run action: no
--------------
00:00:42.004 [ERROR] [../src/input/cursor.c:1296] process_cursor_axis() with start scroll_offset 0.000000, 6.250000 and delta 3.750000, delta discrete 30.000000
00:00:42.004 [ERROR] [../src/input/cursor.c:1322] process_cursor_axis() end scroll_offset 0.000000, 10.000000, delta 3.750000, delta discrete 30.000000 which should run action: no
--------------
```
</details>

I also tried using the actual workspace switcher and it does feel much better, but still a bit “weird” sometimes: Sometimes it switches two workspace for a single click, sometimes a click doesn’t register at all.

(I really appreciate you taking the time to debug this. 🙏)

--- a/include/labwc.h
+++ b/include/labwc.h
@@ -111,9 +111,10 @@ struct seat {
 	bool cursor_visible;
 	struct wlr_cursor *cursor;
 	struct wlr_xcursor_manager *xcursor_manager;
-	struct {
-		double x, y;
-	} smooth_scroll_offset;
+	struct accumulated_scroll {
+		double delta;
+		double delta_discrete;
+	} accumulated_scrolls[2]; /* indexed by wl_pointer_axis */
 	bool cursor_scroll_wheel_emulation;
 
 	/*

--- a/src/input/cursor.c
+++ b/src/input/cursor.c
@@ -1245,38 +1245,44 @@ struct scroll_info {
 };
 
 static struct scroll_info
-compare_delta(double delta, double delta_discrete, double *accum)
+compare_delta(double delta, double delta_discrete, struct accumulated_scroll *accum)
 {
-	/*
-	 * Smooth scroll deltas are in surface space, so treating each unit as a
-	 * scroll event would result in too-fast scrolling.
-	 *
-	 * This fudge factor (inherited from various historic projects, incl. Weston)
-	 * produces events at a more reasonable rate.
-	 *
-	 * For historic context, see:
-	 * https://lists.freedesktop.org/archives/wayland-devel/2019-April/040377.html
-	 */
-	const double SCROLL_THRESHOLD = 10.0;
 	struct scroll_info info = {0};
 
-	if (delta_discrete < 0 || delta < 0) {
-		info.direction = -1;
-	} else if (delta_discrete > 0 || delta > 0) {
-		info.direction = 1;
-	}
-
-	if (delta == 0.0) {
-		/* Delta 0 marks the end of a scroll */
-		*accum = 0.0;
+	if (delta_discrete) {
+		/* mice */
+		info.direction = delta_discrete > 0 ? 1 : -1;
+		accum->delta_discrete += delta_discrete;
+		/*
+		 * Non-hi-res mice produce delta_discrete of ±120 for every
+		 * "click", so it always triggers actions. But for hi-res mice
+		 * that produce smaller delta_discrete, we accumulate it and
+		 * run actions after it exceeds 120(= 1 click).
+		 */
+		if (fabs(accum->delta_discrete) >= 120.0) {
+			accum->delta_discrete = fmod(accum->delta_discrete, 120.0);
+			info.run_action = true;
+		}
 	} else {
-		/* Accumulate smooth scrolling until we hit threshold */
-		*accum += delta;
-	}
-
-	if (delta_discrete != 0 || fabs(*accum) > SCROLL_THRESHOLD) {
-		*accum = fmod(*accum, SCROLL_THRESHOLD);
-		info.run_action = true;
+		/* 2-finger scrolling on touchpads */
+		if (delta == 0) {
+			/* delta=0 marks the end of a scroll */
+			accum->delta = 0;
+			return info;
+		}
+		info.direction = delta > 0 ? 1 : -1;
+		accum->delta += delta;
+		/*
+		 * The threshold of 10 is inherited from various historic
+		 * projects including weston.
+		 *
+		 * For historic context, see:
+		 * https://lists.freedesktop.org/archives/wayland-devel/2019-April/040377.html
+		 */
+		if (fabs(accum->delta) >= 10.0) {
+			accum->delta = fmod(accum->delta, 10.0);
+			info.run_action = true;
+		}
 	}
 
 	return info;
@@ -1290,21 +1296,16 @@ process_cursor_axis(struct server *server, enum wl_pointer_axis orientation,
 	uint32_t modifiers = keyboard_get_all_modifiers(&server->seat);
 
 	enum direction direction = LAB_DIRECTION_INVALID;
-	struct scroll_info info = {0};
+	struct scroll_info info = compare_delta(delta, delta_discrete,
+		&server->seat.accumulated_scrolls[orientation]);
 
 	if (orientation == WL_POINTER_AXIS_HORIZONTAL_SCROLL) {
-		info = compare_delta(delta, delta_discrete,
-			&server->seat.smooth_scroll_offset.x);
-
 		if (info.direction < 0) {
 			direction = LAB_DIRECTION_LEFT;
 		} else if (info.direction > 0) {
 			direction = LAB_DIRECTION_RIGHT;
 		}
 	} else if (orientation == WL_POINTER_AXIS_VERTICAL_SCROLL) {
-		info = compare_delta(delta, delta_discrete,
-			&server->seat.smooth_scroll_offset.y);
-
 		if (info.direction < 0) {
 			direction = LAB_DIRECTION_UP;
 		} else if (info.direction > 0) {
@@ -1324,8 +1325,8 @@ process_cursor_axis(struct server *server, enum wl_pointer_axis orientation,
 					&& mousebind->mouse_event == MOUSE_ACTION_SCROLL) {
 				handled = true;
 				/*
-				 * Action may not be executed if the accumulated scroll
-				 * delta on touchpads doesn't exceed the threshold
+				 * Action may not be executed if the accumulated scroll delta
+				 * on touchpads or hi-res mice doesn't exceed the threshold
 				 */
 				if (info.run_action) {
 					actions_run(ctx.view, server, &mousebind->actions, &ctx);

